overwatch-rs/tests/generics.rs

```diff
@@ -2,51 +2,36 @@ use async_trait::async_trait;
 use futures::future::select;
 use overwatch_derive::Services;
 use overwatch_rs::overwatch::OverwatchRunner;
-use overwatch_rs::services::handle::{ServiceHandle, ServiceStateHandle};
-use overwatch_rs::services::relay::RelayMessage;
+use overwatch_rs::services::handle::ServiceStateHandle;
 use overwatch_rs::services::state::{NoOperator, NoState};
 use overwatch_rs::services::{ServiceCore, ServiceData, ServiceId};
+use overwatch_rs::OpaqueServiceHandle;
 use std::fmt::Debug;
 use std::time::Duration;
 use tokio::time::sleep;
 
-pub struct GenericService<T: Send>
-where
-    T: Debug + 'static + Sync,
-{
-    state: ServiceStateHandle<Self>,
-    _phantom: std::marker::PhantomData<T>,
+pub struct GenericService {
+    state: ServiceStateHandle<GenericServiceMessage, (), Self, NoState<()>>,
 }
 
 #[derive(Clone, Debug)]
 pub struct GenericServiceMessage(String);
 
-impl RelayMessage for GenericServiceMessage {}
-
-impl<T: Send> ServiceData for GenericService<T>
-where
-    T: Debug + 'static + Sync,
-{
+impl ServiceData for GenericService {
     const SERVICE_ID: ServiceId = "FooService";
     type Settings = ();
     type State = NoState<Self::Settings>;
-    type StateOperator = NoOperator<Self::State>;
+    type StateOperator = NoOperator<Self::State, Self::Settings>;
     type Message = GenericServiceMessage;
 }
 
 #[async_trait]
-impl<T: Send> ServiceCore for GenericService<T>
-where
-    T: Debug + 'static + Sync,
-{
+impl ServiceCore for GenericService {
     fn init(
-        state: ServiceStateHandle<Self>,
+        state: ServiceStateHandle<Self::Message, Self::Settings, Self, Self::State>,
```

Comment:
I guess everything has a tradeoff :thinking:

Reply:
Yes, but this only happens in functions declarations. Also, we now provide an opaque version of this type, so we could also replace this with `OpaqueServiceStateHandle<Self>`, and it would still compile.